syz-manager/filter.go

```diff
@@ -0,0 +1,246 @@
+		mgr.kcovFilter.enableFilter = false
+		return
+	}
+	mgr.kcovFilter.enableFilter = true
+	log.Logf(0, "initialize coverage information...")
+	if err := initCover(mgr.sysTarget, mgr.cfg.KernelObj, mgr.cfg.KernelSrc, mgr.cfg.KernelBuildSrc); err != nil {
+		log.Logf(0, "failed to generate coverage profile: %v", err)
+		log.Fatalf("coverage filter cannot be initialized without coverage profile")
+	}
+
+	mgr.kcovFilter.targetLittleEndian = mgr.sysTarget.LittleEndian
+	mgr.kcovFilter.targetArch = mgr.sysTarget.Arch
+	mgr.kcovFilter.pcsBitmapPath = mgr.cfg.Workdir + "/" + "syz-cover-bitmap"
+	mgr.kcovFilter.initWeightedPCs(files, funcs, rawPCs)
+}
+
+func (mgr *Manager) getWeightedPCs() bool {
+	return mgr.kcovFilter.enableFilter
+}
+
+func (filter *CoverFilter) initWeightedPCs(files, functions, rawPCsFiles []string) {
+	filter.weightedPCs = make(map[uint32]float32)
+	filesRegexp := getFileRegexp(files)
+	funcsRegexp := getFuncRegexp(functions)
+
+	enabledFiles := make(map[string]bool)
+	enabledFuncs := make(map[string]bool)
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			log.Logf(0, "failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			filter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+	pcs := reportGenerator.PCs()
+	for _, e := range pcs {
+		frame := e[len(e)-1]
+		fullpc := cover.NextInstructionPC(filter.targetArch, frame.PC)
+		pc := uint32(fullpc & 0xffffffff)
+		for _, r := range funcsRegexp {
+			if ok := r.MatchString(frame.Func); ok {
+				enabledFuncs[frame.Func] = true
+				filter.weightedPCs[pc] = 1.0
+			}
+		}
+		for _, r := range filesRegexp {
+			if ok := r.MatchString(frame.File); ok {
+				enabledFiles[frame.File] = true
+				enabledFuncs[frame.Func] = true
+				filter.weightedPCs[pc] = 1.0
+			}
+		}
+		if _, ok := filter.weightedPCs[pc]; ok {
+			enabledFuncs[frame.Func] = true
+		}
+	}
+	for f := range enabledFuncs {
+		log.Logf(1, "enabled func: %s", f)
+	}
+	for f := range enabledFiles {
+		log.Logf(1, "enabled file: %s", f)
+	}
+}
+
+func getFileRegexp(files []string) []regexp.Regexp {
+	var regexps []regexp.Regexp
+	// `\*{2}$` match net/dccp/**
+	rMatchDoubleStart, err := regexp.Compile(`\*{2}$`)
+	if err != nil {
+		log.Fatalf("regular expression failed: %s", err)
+	}
+	// `[^\\*]\\*$` match net/sctp/*
+	rMatchOneStart, err := regexp.Compile(`[^\*]\*$`)
+	if err != nil {
+		log.Fatalf("regular expression failed: %s", err)
+	}
+	for _, f := range files {
+		if ok1 := rMatchDoubleStart.MatchString(f); ok1 {
+			f = `^` + f[:len(f)-2]
+		} else if ok2 := rMatchOneStart.MatchString(f); ok2 {
+			f = `^` + f[:len(f)-1] + `[^\/]*$`
+		} else {
+			f = f + `$`
+		}
+		r, err := regexp.Compile(f)
+		if err != nil {
+			log.Fatalf("regular expression failed: %s", err)
```

Comment:
Code (except for main function) should not call log.Fatalf, but instead return errors. This allows better error handling and provides testability. Make all functions return error instead and add a single log.Fatalf at the caller side.